src/main/groovy/io/openliberty/tools/gradle/tasks/DevTask.groovy

```diff
@@ -0,0 +1,507 @@
+                File defaultResourceDir = new File(project.getRootDir() + "/src/main/resources");
+                logger.info("No resource directory detected, using default directory: " + defaultResourceDir);
+                resourceDirs.add(defaultResourceDir);
+            }
+
+            String artifactId = project.getName();
+
+            // create an executor for tests with an additional queue of size 1, so
+            // any further changes detected mid-test will be in the following run
+            final ThreadPoolExecutor executor = new ThreadPoolExecutor(1, 1, 0L, TimeUnit.MILLISECONDS,
+                    new ArrayBlockingQueue<Runnable>(1, true));
+
+            ProjectConnection gradleConnection = initGradleProjectConnection();
+            BuildLauncher gradleBuildLauncher = gradleConnection.newBuild();
+            try {
+                /*
+                Running the installApps task runs all tasks it depends on:
+                    :libertyStop
+                    :clean
+                    :installLiberty
+                    :libertyCreate
+                    :compileJava
+                    :processResources
+                    :classes
+                    :war
+                    :installApps
+                 */
+                runGradleTask(gradleBuildLauncher, 'libertyCreate');
+                runGradleTask(gradleBuildLauncher, 'installFeature');
+                runGradleTask(gradleBuildLauncher, 'installApps');
+            } finally {
+                gradleConnection.close();
+            }
+
+            util = new DevTaskUtil(
+                    serverDirectory, sourceDirectory, testSourceDirectory, configDirectory,
+                    resourceDirs, hotTests, skipTests, artifactId, serverStartTimeout,
+                    verifyTimeout, appUpdateTimeout, compileWait, libertyDebug
+            );
+
+//            Use the gradle compile task instead of using the DevUtil compile
+//            util.setUseMavenOrGradleCompile(true);
+
+            util.addShutdownHook(executor);
+
+            util.startServer();
+
+            List<String> artifactPaths = util.getArtifacts();
+            File buildFile = project.getBuildFile();
+            File serverXMLFile = getServerXMLFile(server);
+
+            if (hotTests && testSourceDirectory.exists()) {
+            // if hot testing, run tests on startup and then watch for keypresses
+            util.runTestThread(false, executor, -1, false, false);
+            } else {
+                // else watch for key presses immediately
+                util.runHotkeyReaderThread(executor);
+            }
+
+            try {
+                util.watchFiles(buildFile, outputDirectory, testOutputDirectory, executor, artifactPaths, serverXMLFile);
+            } catch (PluginScenarioException e) { // this exception is caught when the server has been stopped by another process
+                logger.info(e.getMessage());
```

Comment:
Shouldn't we provide some context here for this log message?

Reply:
@cherylking This message comes from the exception thrown by ci.common's DevUtil.java which already provides the context: `throw new PluginScenarioException("The server has stopped. Exiting dev mode.");`